### PR TITLE
Replace internal PublicKey type with one from proof-systems

### PR DESCRIPTION
**Pull request by willemolding**
# Replace internal PublicKey type with one from proof-systems

This PR is a big tidy up. It removes one placeholder types and tidies up a bunch of derives on our internal types. Apologies for the many small line changes (most are just removing derives).

Also note this references a fork of `proof-systems` that adds some helpful derives to their `CompressedPubKey` type. I've submitted a PR back to the upstream to add it. Without these derives most of our tests can't work. 

**Summary of changes**
Changes introduced in this pull request:
- Remove our placeholder `PublicKey` type
- Add conversions from the serialization types `PublicKeyV1` and `PublicKey2V1` to the proof-systems type `CompressedPubKey` 
- Remove unnecessary derives of `Serialize`, `Deserialize` and `Default` from internal types. These are actually potentially harmful as:
    - some types don't have sensible defaults (e.g what does a default public key mean?) 
    - serialization should go through the serialization types or the results will be wrong
    - Also, CompressedPubKey does not implement these derives so any containing types can't automatically derive them
- Update some tests to make less use of default
- Update the `UserCommandBuilder` to also not make use of defaults
- Remove a flaky test in `consensus/tests/common.rs` that depended on the placeholder hashing algorithm (which changed when the data structures changed)

**Reference issue to close (if applicable)**

Progress on #193 (PubKey only, still need to do Signature)

**Other information and links**

### CODE REVIEW

File: consensus/tests/common.rs [hunk 1]
```diff
@@ -127,76 +127,83 @@ mod tests {
...
-        assert_eq!(
-            canonical.body.consensus_state.min_window_density,
-            Length(43)
-        );
-        assert_eq!(
-            canonical.body.consensus_state.sub_window_densities,
-            densities
-        );
-    }
+    // @willemolding:
```
Comment (willemolding) [inline consensus/tests/common.rs:130]: It would be great to open a discussion around this

File: protocol/serialization-types/src/signatures.rs [hunk 2]
```diff
@@ -21,10 +21,50 @@ pub struct CompressedCurvePoint {
...
 
 /// Public key (v1) with an extra version byte
-pub type PublicKey2V1 = Versioned<PublicKeyV1, 1>; // with an extra version wrapper
+#[derive(Clone, Debug, Serialize, Deserialize, PartialEq)]
+pub struct PublicKey2V1(pub Versioned<PublicKeyV1, 1>); // with an extra version wrapper
 
 /// Signature (v1)
 pub type SignatureV1 = Versioned<Versioned<(FieldElement, InnerCurveScalar), 1>, 1>;
+
+mod conversions {
```
Comment (willemolding) [inline protocol/serialization-types/src/signatures.rs:34]: This is the main conversion code to be reviewed

File: base/src/consensus_state.rs [hunk 3]
```diff
@@ -66,11 +68,14 @@ pub struct ConsensusState {
     /// If the block has an ancestor in the same checkpoint window
     pub has_ancestor_in_same_checkpoint_window: bool,
     /// Compressed public key of winning account
-    pub block_stake_winner: PublicKey,
+    #[default(CompressedPubKey::from_address("B62qiy32p8kAKnny8ZFwoMhYpBppM1DWVCqAPBYNcXnsAHhnfAAuXgg").unwrap())]
```
Comment (hanabi1224) [inline base/src/consensus_state.rs:71]: Sweet!

File: base/src/user_commands/signed_command.rs [hunk 4]
```diff
@@ -4,25 +4,26 @@
 //! Signed commands are commands that require signing with some accounts private key
 
 use crate::numbers::{Amount, ExtendedU32, TokenId};
-use mina_crypto::signature::{PublicKey, Signature};
-use serde::{Deserialize, Serialize};
+use mina_crypto::signature::Signature;
+
```
Comment (hanabi1224) [inline base/src/user_commands/signed_command.rs:8]: Nit: unwanted spaces?

File: consensus/Cargo.toml [hunk 5]
```diff
@@ -21,6 +21,8 @@ hex = "0.4"
 lazy_static = "1"
 thiserror = "1"
 
+mina-signer = {git = "https://github.com/willemolding/proof-systems", rev = "c05889b53668b44bb8944737581a24399747fb78"}
```
Comment (hanabi1224) [inline consensus/Cargo.toml:24]: is this gonna merged?

File: protocol/serialization-types/src/signatures.rs [hunk 6]
```diff
@@ -21,10 +21,50 @@ pub struct CompressedCurvePoint {
...
+    use super::{CompressedCurvePoint, PublicKey2V1, PublicKeyV1};
+    use mina_signer::{BaseField, CompressedPubKey};
+    use o1_utils::field_helpers::FieldHelpers;
+    use versioned::Versioned;
+
+    impl From<PublicKeyV1> for CompressedPubKey {
+        fn from(t: PublicKeyV1) -> Self {
+            CompressedPubKey {
+                // This unwrap is safe as a PublicKeyV1 always has 32 bytes of data and from_bytes does not check if it is on curve
+                x: BaseField::from_bytes(&t.0.t.t.x).unwrap(),
```
Comment (hanabi1224) [inline protocol/serialization-types/src/signatures.rs:44]: maybe expect with verbose error message?

**Review by hanabi1224 — APPROVED** [review]

File: consensus/Cargo.toml (hunk 5 above)
Comment (willemolding) [reply consensus/Cargo.toml:24]: Hopefully. We should hold off merging this and see. It isn't ideal to be referencing a fork on my github account

File: proof-systems-shim/Cargo.toml [hunk 7]
```diff
@@ -7,5 +7,11 @@ edition = "2021"
...
 
 [dependencies]
-commitment_dlog = {git = "https://github.com/o1-labs/proof-systems", rev = "56517b55920ea7afc811baa541c859dadc8ae50a"}
 mina-hasher = {git = "https://github.com/o1-labs/proof-systems", rev = "56517b55920ea7afc811baa541c859dadc8ae50a"}
+commitment_dlog = {git = "https://github.com/o1-labs/proof-systems", rev = "6af257b793d80f900675acc01b16fd4bf6375268"}
+mina-curves = {git = "https://github.com/o1-labs/proof-systems", rev = "6af257b793d80f900675acc01b16fd4bf6375268"}
+plonk_circuits = {git = "https://github.com/o1-labs/proof-systems", rev = "6af257b793d80f900675acc01b16fd4bf6375268"}
+plonk_protocol_dlog = {git = "https://github.com/o1-labs/proof-systems", rev = "6af257b793d80f900675acc01b16fd4bf6375268"}
+
+mina-signer = {git = "https://github.com/o1-labs/proof-systems", rev = "1f88a94eddba96b720c9214303fb2b4b0cd97597"}
```
Comment (hanabi1224) [inline proof-systems-shim/Cargo.toml:16]: Should we keep the rev(s) identical here?

File: consensus/tests/common.rs (hunk 1 above)
Comment (creativcoder) [reply consensus/tests/common.rs:130]: thanks, will take a look at this in a follow-up PR.

**Review by creativcoder — APPROVED** [review]
Thanks for this refactor :rescue_worker_helmet:

**Pull request merged**